crates/fuel-core/src/database.rs

```diff
@@ -491,6 +289,176 @@ impl AtomicView for Database<Relayer> {
+            ))
+            .map(|result| result.map(|(height, _)| height))
+            .try_collect()
+        })
+    }
+}
+
+#[cfg(not(feature = "relayer"))]
+impl Modifiable for Database<Relayer> {
+    fn commit_changes(&mut self, changes: Changes) -> StorageResult<()> {
+        commit_changes_with_height_update(self, changes, |_| Ok(vec![]))
+    }
+}
+
+trait DatabaseHeight: Sized {
+    fn as_u64(&self) -> u64;
+
+    fn advance_height(&self) -> Option<Self>;
+}
+
+impl DatabaseHeight for BlockHeight {
+    fn as_u64(&self) -> u64 {
+        let height: u32 = (*self).into();
+        height as u64
+    }
+
+    fn advance_height(&self) -> Option<Self> {
+        self.succ()
+    }
+}
+
+impl DatabaseHeight for DaBlockHeight {
+    fn as_u64(&self) -> u64 {
+        self.0
+    }
+
+    fn advance_height(&self) -> Option<Self> {
+        self.0.checked_add(1).map(Into::into)
+    }
+}
+
+fn commit_changes_with_height_update<Description>(
+    database: &mut Database<Description>,
+    changes: Changes,
+    heights_lookup: impl Fn(
+        &ChangesIterator<Description>,
+    ) -> StorageResult<Vec<Description::Height>>,
+) -> StorageResult<()>
+where
+    Description: DatabaseDescription,
+    Description::Height: Debug + PartialOrd + DatabaseHeight,
+    for<'a> StorageTransaction<&'a &'a mut Database<Description>>:
+        StorageMutate<MetadataTable<Description>, Error = StorageError>,
+{
+    // Gets the all new heights from the `changes`
+    let iterator = ChangesIterator::<Description>::new(&changes);
+    let new_heights = heights_lookup(&iterator)?;
+
+    // Changes for each block should be committed separately.
+    // If we have more than one height, it means we are mixing commits
+    // for several heights in one butch - return error in this case.
+    if new_heights.len() > 1 {
+        return Err(DatabaseError::MultipleHeightsInCommit {
+            heights: new_heights.iter().map(DatabaseHeight::as_u64).collect(),
+        }
+        .into());
+    }
+
+    let new_height = new_heights.into_iter().last();
+    let prev_height = *database.height.lock();
+
+    match (prev_height, new_height) {
+        (None, None) => {
+            // We are inside the regenesis process if the old and new heights are not set.
+            // In this case, we continue to commit until we discover a new height.
+            // This height will be the start of the database.
+        }
+        (Some(prev_height), Some(new_height)) => {
+            // Each new commit should be linked to the previous commit to create a monotonically growing database.
+
+            let next_expected_height = prev_height
+                .advance_height()
+                .ok_or(DatabaseError::FailedToAdvanceHeight)?;
+
+            // TODO: After https://github.com/FuelLabs/fuel-core/issues/451
+            //  we can replace `next_expected_height > new_height` with `next_expected_height != new_height`.
+            if next_expected_height > new_height {
```

Comment:
Is there a specific reason that using `!=` depends on issue 451? 
It looks like this is saying that `next_expected_height` is allowed to be equal to or less than `new_height`.

Reply:
Right now, the relayer commits only when there are messages. If the DA block has no messages, the relayer doesn't commit.